airflow/creds.py

```diff
@@ -0,0 +1,171 @@
+    """
+    Retrieves Connection object from airflow metastore database.
+    """
+
+    # pylint: disable=missing-docstring
+    @provide_session
+    def get_connections(self, conn_id, session=None) -> List[Connection]:
+        conn_list = (
+            session.query(Connection).filter(Connection.conn_id == conn_id).all()
+        )
+        session.expunge_all()
+        return conn_list
+
+
+class AwsSsmCredsBackend(BaseCredsBackend):
+    """
+    Retrieves Connection object from AWS SSM Parameter Store
+
+    Configurable via airflow config like so:
+
+    .. code-block:: ini
+
+        [aws_ssm_creds]
+        ssm_prefix = /airflow
+        profile_name = default
+
+    For example, if ssm path is ``/airflow/AIRFLOW_CONN_SMTP_DEFAULT``, this would be accessible if you
+    provide ``ssm_prefix = /airflow`` and conn_id ``smtp_default``.
+
+    """
+
+    conf_section = "aws_ssm_creds"
+    conf_key_ssm_prefix = "ssm_prefix"
+    default_prefix = "/airflow"
+    conf_key_profile_name = "profile_name"
+    default_profile = "default"
+
+    def __init__(self, *args, **kwargs):
+        pass
+
+    @property
+    def ssm_prefix(self) -> str:
+        """
+        Gets ssm prefix from conf.
+
+        Ensures that there is no trailing slash.
+
+        :return:
+        """
+        ssm_prefix = conf.get(
+            section=self.conf_section,
+            key=self.conf_key_ssm_prefix,
+            fallback=self.default_prefix,
+        )
+        if ssm_prefix[-1] == "/":
+            ssm_prefix = ssm_prefix[:-1]
+        return ssm_prefix
+
+    @property
+    def aws_profile_name(self) -> str:
+        """
+        Gets AWS profile to use from conf.
+
+        :return:
+        """
+        profile_name = conf.get(
+            section=self.conf_section,
+            key=self.conf_key_profile_name,
+            fallback=self.default_profile,
+        )
+        return profile_name
+
+    def build_ssm_path(self, conn_id: str):
+        """
+        Given conn_id, build SSM path.
+
+        Assumes connection params use same naming convention as env vars, but may have arbitrary prefix.
+
+        :param conn_id:
+        :return:
+        """
+        param_name = (CONN_ENV_PREFIX + conn_id).upper()
+        param_path = self.ssm_prefix + "/" + param_name
+        return param_path
+
+    def get_conn_uri(self, conn_id):
+        """
+        Get param value
+
+        :param conn_id:
+        :return:
+        """
+        session = boto3.Session(profile_name=self.aws_profile_name)
+        client = session.client("ssm")
+        response = client.get_parameter(
+            Name=self.build_ssm_path(conn_id=conn_id), WithDecryption=True
+        )
+        value = response["Parameter"]["Value"]
+        return value
+
+    def get_connections(self, conn_id) -> List[Connection]:
+        """
+        Create connection object.
+
+        :param conn_id:
+        :return:
+        """
+        conn_uri = self.get_conn_uri(conn_id=conn_id)
```

Comment:
Storing connections in a URL is probably not the best idea. URL support is related to a long history of errors that I solved. Maybe it is better to serialize the object to another format, e.g. JSON?

Reply:
I haven't encountered any problem with URI format so it did not occur to me to change it.  It's used currently for both env var conn def and `airflow connections -a` so it seemed natural to keep it.

But I like your idea.  If we are storing elsewhere, and building separate "backend", we don't have to adhere to same format.

Here's an argument for keeping with the URI format.  Keeping the same URI format means that it would be easier to change from one creds provider to another.  If you are using env vars now, it would be extremely easy to switch to SSM, or vice versa if you want to switch back, or override something locally.

But maybe this question should be considered separately, in PRs for individual providers e.g. AWS SSM or whatever.

I suppose the first step is to agree on how to structure the backend framework (e.g. what is the right base class, or what functions should be in a creds provider implementation).  Then the individual details for specific creds providers could be handled separately, or implemented arbitrarily in accordance with the preferences of individual users.  

Do you think it's a good idea for me to remove the SSM piece from this PR so this is more focused on the creds provider abstraction?